File: windows/src/Patchthrough.Core/VocabularyEvidence.cs
namespace Patchthrough.Core;

public static class VocabularyEvidence
{
    public static IReadOnlyList<string> AcousticallySupported(
        IEnumerable<string> requested,
        IReadOnlyList<EngineWord> words,
        double minimumConfidence = 0.72)
    {
        var normalized = words.Select(word => Normalize(word.Text)).ToList();
        return requested.Where(term =>
        {
            var parts = term.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)
                .Select(Normalize).Where(part => part.Length > 0).ToList();
            if (parts.Count == 0 || normalized.Count < parts.Count) return false;
            for (var start = 0; start <= normalized.Count - parts.Count; start++)
            {
                if (!normalized.Skip(start).Take(parts.Count).SequenceEqual(parts)) continue;
                var confidence = words.Skip(start).Take(parts.Count).Select(word => word.Confidence).ToList();
                if (confidence.All(value => value.HasValue)
                    && confidence.Average(value => value!.Value) >= minimumConfidence) return true;
            }
            return false;
        }).ToList();
    }

    private static string Normalize(string value) =>
        new(value.ToLowerInvariant().Where(char.IsLetterOrDigit).ToArray());
}

File: Sources/patchthrough/Transcription/VocabularyEvidence.swift
import Foundation

enum VocabularyEvidence {
    static func acousticallySupported(
        _ requested: [String],
        in words: [TimedWord],
        minimumConfidence: Double = 0.72
    ) -> [String] {
        let normalizedWords = words.map { normalize($0.text) }
        return requested.filter { term in
            let parts = term.split(whereSeparator: \.isWhitespace).map { normalize(String($0)) }.filter { !$0.isEmpty }
            guard !parts.isEmpty, normalizedWords.count >= parts.count else { return false }
            for start in 0...(normalizedWords.count - parts.count) {
                guard Array(normalizedWords[start..<(start + parts.count)]) == parts else { continue }
                let confidence = words[start..<(start + parts.count)].compactMap(\.confidence)
                if confidence.count == parts.count && confidence.reduce(0, +) / Double(confidence.count) >= minimumConfidence {
                    return true
                }
            }
            return false
        }
    }

    private static func normalize(_ value: String) -> String {
        value.lowercased().filter { $0.isLetter || $0.isNumber }
    }
}
